packages/@aws-cdk/aws-codestarnotifications/lib/notification-rule.ts

```diff
@@ -0,0 +1,237 @@
+  /**
+   * A list of event types associated with this notification rule.
+   * For a complete list of event types and IDs, see Notification concepts in the Developer Tools Console User Guide.
+   * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#concepts-api
+   *
+   * You must specify this property (either via props or via `addEvents`)
+   *
+   * @default - No events.
+   */
+  readonly events?: string[];
+}
+
+/**
+ * Properties for a new notification rule
+ */
+export interface NotificationRuleProps extends NotifyOnEventOptions {
+  /**
+   * The Amazon Resource Name (ARN) of the resource to associate with the notification rule.
+   * Currently, Supported sources include pipelines in AWS CodePipeline and build projects in AWS CodeBuild in this L2 constructor.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codestarnotifications-notificationrule.html#cfn-codestarnotifications-notificationrule-resource
+   */
+  readonly source: INotificationRuleSource;
+}
+
+/**
+ * Represents a notification rule
+ */
+export interface INotificationRule extends IResource {
+
+  /**
+   * The ARN of the notification rule (i.e. arn:aws:codestar-notifications:::notificationrule/01234abcde)
+   *
+   * @attribute
+   */
+  readonly notificationRuleArn: string;
+
+  /**
+   * Adds target to notification rule
+   *
+   * @param target The SNS topic or AWS Chatbot Slack target
+   * @returns boolean - return true if it had any effect
+   */
+  addTarget(target?: INotificationRuleTarget): boolean;
+}
+
+/**
+ * A new notification rule
+ *
+ * @resource AWS::CodeStarNotifications::NotificationRule
+ */
+export class NotificationRule extends Resource implements INotificationRule {
+  /**
+   * Import an existing notification rule provided an ARN
+   * @param scope The parent creating construct
+   * @param id The construct's name
+   * @param notificationRuleArn Notification rule ARN (i.e. arn:aws:codestar-notifications:::notificationrule/01234abcde)
+   */
+  public static fromNotificationRuleArn(scope: Construct, id: string, notificationRuleArn: string): INotificationRule {
+    class Import extends Resource implements INotificationRule {
+      readonly notificationRuleArn = notificationRuleArn;
+
+      public addTarget(_target?: INotificationRuleTarget) {
+        return false;
+      }
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * @attribute
+   */
+  readonly notificationRuleArn: string;
+
+  /**
+   * The source config of notification rule
+   */
+  readonly source: NotificationRuleSourceConfig;
+
+  /**
+   * The target config of notification rule
+   */
+  readonly targets: NotificationRuleTargetConfig[] = [];
+
+  /**
+   * The events of notification rule
+   */
+  readonly events: string[] = [];
+
+  constructor(scope: Construct, id: string, props: NotificationRuleProps) {
+    super(scope, id);
+
+    this.source = props.source.bind(this);
```

Comment:
Notice that you're passing `this` before you assign any of its properties to it. This is bad; that means if anyone inside `bind()` calls `rule.notificationRuleArn`, they will get undefined.

Because of that, change `bind()` in `INotificationRuleSource` to take a `scope: Construct` instead of an `INotificationRule` as its argument.

Reply:
Got it, I will adjust.